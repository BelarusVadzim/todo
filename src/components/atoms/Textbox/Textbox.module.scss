@import 'src/variables.scss';

.textbox {
  font-size: $font-size;
  color: $primary-font-color;
  font-family: $main-font;
  background-color:  transparent;
  caret-color: $primary-color;
  outline: none;
  border: none;
  // border-bottom: 1px solid transparent;
  // animation-name: blinking;
  // animation-duration: 2s;
  // animation-iteration-count: 3;

  // &:hover {
  //   border-bottom: 1px solid $primary-color;
  //   }

  // @keyframes blinking {
  //   50% {
  //     border-color: $primary-color;
  //   }
  // }
}
