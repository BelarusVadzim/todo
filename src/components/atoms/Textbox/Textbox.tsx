import React from 'react';
import style from './Textbox.module.scss';
import { useCallback, useState } from 'react';

type TextboxProps = {
  onReturn?: (text: string) => void;
};

const EnterKey = 'Enter';

const Textbox: React.FC<TextboxProps> = ({ onReturn }) => {
  const [text, setText] = useState('');

  const keyDownHandler = useCallback(
    (key: string, value: string) => {
      if (EnterKey === key) {
        if ( onReturn ) onReturn(value);
        setText('');
      }
    },
    [onReturn],
  );

  return (
    <input
      type="text"
      className={style.textbox}
      onChange={(ChangeEventArg) => setText(ChangeEventArg.target.value)}
      onKeyDown={(keyboardEventArg) =>
        keyDownHandler(
          keyboardEventArg.key,
          keyboardEventArg.currentTarget.value,
        )
      }
      value={text}
      maxLength={50}
    />
  );
};

export default Textbox;
