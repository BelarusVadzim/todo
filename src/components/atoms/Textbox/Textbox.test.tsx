import render  from 'react-test-renderer';
import Textbox from './Textbox';

const textValue = 'textValue';
jest.mock('./Textbox.module.scss', () => ({ textbox: 'textboxStyle' }));
let mockSetText: jest.Mock;
jest.mock('react', () => {
  mockSetText = jest.fn();
  const actual = jest.requireActual('react');
  return {
    ...actual,
    useState: jest.fn().mockReturnValue([textValue, mockSetText]),
  };
});

afterEach(() => {
  jest.clearAllMocks();
});

describe('<Textbox />', () => {
  it('should render properly', () => {
    const component = render
      .create(<Textbox />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('when onKeyDown raised witn not Enter key should call expected functions with expected values', () => {
    const onReturnHandler = jest.fn();
    const value = 'textbox value';
    const key = 'notEnter';
    const component = render
      .create(<Textbox 
        onReturn={onReturnHandler} />);

    const keyboardEventArg = { key, currentTarget: { value } };

    component.root.findByType('input').props.onKeyDown(keyboardEventArg);

    expect(onReturnHandler).toBeCalledTimes(0);
    expect(mockSetText).toBeCalledTimes(0);
  });

  it('when onKeyDown raised witn Enter key should call expected functions with expected values', () => {
    const onReturnHandler = jest.fn();
    const value = 'textbox value';
    const key = 'Enter';
    const component = render
      .create(<Textbox 
        onReturn={onReturnHandler} />);

    const expectedValue = value;
    const keyboardEventArg = { key, currentTarget: { value } };

    component.root.findByType('input').props.onKeyDown(keyboardEventArg);

    expect(onReturnHandler).toBeCalledTimes(1);
    expect(onReturnHandler).toBeCalledWith(expectedValue);
    expect(mockSetText).toBeCalledTimes(1);
    expect(mockSetText).toBeCalledWith('');
    
  });

  it('when onChange raised should call expected functions with expected values', () => {
    const value = 'textbox value';
    const component = render
      .create(<Textbox />);

    const expectedValue = value;
    const changeEventArg = { target: { value } };

    component.root.findByType('input').props.onChange(changeEventArg);

    expect(mockSetText).toBeCalledTimes(1);
    expect(mockSetText).toBeCalledWith(expectedValue);
    
  });
});
