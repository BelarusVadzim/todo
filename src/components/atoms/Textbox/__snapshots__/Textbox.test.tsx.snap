// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Textbox /> should render properly 1`] = `
<input
  className="textboxStyle"
  maxLength={50}
  onChange={[Function]}
  onKeyDown={[Function]}
  type="text"
/>
`;
